ethereum_history_api/circuits/lib/src/rlp/decode.nr

```diff
@@ -56,32 +51,63 @@ pub fn decode_to_rlp_header<N>(arr: [u8; N]) -> RlpHeader {
         RlpHeader { offset, length, data_type: LIST }
     } else { // >55-byte list
         let offset = wrapping_sub(1 + prefix, 0xf7) as u64;
-        let length = extract_payload_len(arr, wrapping_sub(prefix, 0xf7) as u64);
+        let length = extract_payload_len(data, wrapping_sub(prefix, 0xf7) as u64);
         assert(offset + length <= N, "Decoded length of long list exceeds input length");
 
         RlpHeader { offset, length, data_type: LIST }
     }
 }
 
-pub fn decode_string<N>(input: [u8; N]) -> (u64, u64) {
+pub fn decode_string<N>(input: Fragment<N, u8>) -> (u64, u64) {
     let mut RlpHeader {offset, length, data_type} = decode_to_rlp_header(input);
     assert(data_type == STRING, "Wrong data type");
 
     let total_len = length + offset;
-    assert(total_len <= input.len(), "Total length exceeds input length");
+    assert(total_len <= input.length, "Total length exceeds input length");
 
     (offset, length)
 }
 
 /// Returns an RLP list look-up table.
 /// For string elements, the offset points to the payload, whereas the offset
 ///   of a list element points to the RLP header of that element.
-pub fn decode_list<N, NUM_FIELDS>(input: [u8; N]) -> RlpList<NUM_FIELDS> {
+pub fn decode_list_fragments<N, NUM_FIELDS>(input: Fragment<N, u8>) -> RlpList<NUM_FIELDS> {
```

Comment:
`NUM_FIELDS` suggests that it is exact number of rlp list fields but in fact this is maximum number, maybe rename to `MAX_FIELDS`

Reply:
Changed it